lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -591,35 +591,41 @@ void PartitionPass::bubbleUpGlobalRefs(Operation *op) {
   if (!globalRefs)
     return;
 
+  // GlobalRefs use the inner_sym attribute, so keep it up to date.
+  op->setAttr("inner_sym", StringAttr::get(op->getContext(), ::getOpName(op)));
+
   for (auto globalRef : globalRefs.getAsRange<hw::GlobalRefAttr>()) {
     // Resolve the GlobalRefOp and get its path.
     auto refSymbol = globalRef.getGlblSym();
     auto globalRefOp = dyn_cast_or_null<hw::GlobalRefOp>(
         topLevelSyms.getDefinition(refSymbol));
     assert(globalRefOp && "symbol must reference a GlobalRefOp");
-    auto oldPath = globalRefOp.namepath().getAsRange<hw::InnerRefAttr>();
-
-    // Construct a new path starting from the old path.
-    SmallVector<Attribute> newPath(oldPath);
+    auto oldPath = globalRefOp.namepath().getValue();
+    assert(!oldPath.empty());
 
-    // If there aren't two elements in the path, this GlobalRefOp has already
-    // been updated.
-    if (newPath.size() < 2)
+    // If the path already points to the target design partition, we are done.
+    auto leafModule = oldPath.back().cast<hw::InnerRefAttr>().getModule();
+    auto partAttr = op->getAttrOfType<SymbolRefAttr>("targetDesignPartition");
+    if (partAttr.getRootReference() == leafModule)
       return;
+    assert(oldPath.size() > 1);
+
+    // Construct a new path starting from the old path.
+    SmallVector<Attribute> newPath(oldPath.begin(), oldPath.end());
 
     // Splice the last two elements in the path, using the combined name from
     // the op's attributes, and taking the module from the second to last
     // element.
-    newPath.pop_back_val().cast<hw::InnerRefAttr>();
+    auto lastElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
     auto nextElement = newPath.pop_back_val().cast<hw::InnerRefAttr>();
     auto newModule = nextElement.getModule();
-    auto newNameAttr = StringAttr::get(op->getContext(), ::getOpName(op));
+    auto lastName = lastElement.getName().getValue();
+    auto nextName = nextElement.getName().getValue();
+    auto newNameAttr =
+        StringAttr::get(op->getContext(), nextName + "." + lastName);
```

Comment:
Does this have correspond to something elsewhere? I notice you removed the `setAttr("inner_sym", ...)` statement which was setting it.

Reply:
This was a change I made in a previous revision, but we should just be able to use the symbol here. I will update that.

Reply:
So I took another look at this. This does correspond to the `setAttr("inner_sym", ...)` call, which I've moved up to the top above the loop. However, it doesn't work to simply use that inner_sym here. In this loop, for a given instance that has been bubbled up, we go through _all_ the paths that terminate at this instance and update the global refs to what they will be after the `setAttr("inner_sym", ...)` call. It could be possible to change this to only update the specific path that just bubbled up, but this gets tricky for the reasons I mentioned above about keeping the symbols in sync as that happens. So the current approach is about having a reasonably simple implementation that depends more on the iteration order to save on bookkeeping. Does that make sense?